### PR TITLE
Cache metric term annotation inspections

### PR DESCRIPTION
**Pull request by godofwharf**
…ation calls

### CODE REVIEW

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java [hunk 1]
```diff
@@ -90,7 +87,14 @@ private boolean cacheDisabled() {
         return FunctionMetricsManager.getOptions().map(Options::isDisableCacheOptimisation).orElse(false);
     }
 
-    private MethodData createMethodData(Signature callSignature) {
+    private MethodData getMethodData(final ProceedingJoinPoint joinPoint, final Signature callSignature) {
+        return cacheDisabled()
```
Comment (santanusinha) [inline src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:91]: Change cacheDisabled to return true if param capture is enabled irrespective of the other option

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java [hunk 2]
```diff
@@ -99,7 +103,10 @@ private MethodData createMethodData(Signature callSignature) {
         final String methodName = Strings.isNullOrEmpty(monitoredFunction.method())
                                   ? callSignature.getName()
                                   : monitoredFunction.method();
-        return new MethodData(className, methodName);
+        final Options options = FunctionMetricsManager.getOptions().orElse(null);
```
Comment (santanusinha) [inline src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:106]: map to boolean here directly and pass that

File: src/main/java/io/appform/functionmetrics/MethodData.java [hunk 3]
```diff
@@ -22,10 +22,12 @@
 public class MethodData {
     private final String className;
     private final String methodName;
+    private final boolean parameterCaptureRequired;
```
Comment (santanusinha) [inline src/main/java/io/appform/functionmetrics/MethodData.java:25]: You can actually store a list of parameter indexes that need to be read instead of the boolean. As a matter of fact you can store a commons StrSubstitutor format string here and fill in the params at runtime to generate the metric name in a single shot.
Can also add a isParamCaptureRequired() method to return !list.isEmpty()
Advantages:
- SPEEEEEED

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java (hunk 1 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:91]: I don't understand this. Should cache be automatically disabled if param capture is enabled? This needs to be communicated to the user as well. I suggest doing this in the function metrics manager initialisation then.

File: src/main/java/io/appform/functionmetrics/MethodData.java (hunk 3 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/MethodData.java:25]: Ack. Enabling parameter capture definitely has some overhead. As per actual stats in production, even when parameter capture is enabled, it is done so for less than 5% of the methods which are monitored. So I wanted to avoid this overhead of constructing parameter string for methods that don't require it (close to 95%). I will pick this optimisation as a follow up. I will add a TODO for now.

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java (hunk 2 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:106]: Done.

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java [hunk 4]
```diff
@@ -109,20 +116,38 @@ private FunctionInvocation createFunctionInvocation(
 
         final String className = methodData.getClassName();
         final String methodName = methodData.getMethodName();
-        final String parameterString = createParamString(className, methodName, joinPoint, methodSignature, options)
-                .orElse("");
-
+        String parameterString = "";
```
[the remark this answers is not in the document]
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:119]: Done.

File: src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java [hunk 5]
```diff
@@ -54,14 +54,15 @@ public static void initialize(final String packageName, final MetricRegistry reg
...
+        if (null == registry) {
+            log.info("Please call FunctionMetricsManager.initialize() to setup metrics collection. No metrics will be pushed.");
             return Optional.empty();
         }
         MetricRegistry.MetricSupplier<Timer> metricSupplier = () -> new Timer(new SlidingTimeWindowArrayReservoir(60, TimeUnit.SECONDS));
 
-        final String metricName = options.isEnableParameterCapture() && !Strings.isNullOrEmpty(invocation.getParameterString())
-                ? prefix + "." + invocation.getClassName() + "." + invocation.getMethodName() + "." + invocation.getParameterString() + "." + domain.getValue()
+        final String metricName = options.isEnableParameterCapture()
+                && !Strings.isNullOrEmpty(invocation.getParameterString()) ?
```
[the remark this answers is not in the document]
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java:64]: Done.

File: src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java [hunk 6]
```diff
@@ -54,14 +54,15 @@ public static void initialize(final String packageName, final MetricRegistry reg
     }
 
     public static Optional<Timer> timer(final TimerDomain domain, final FunctionInvocation invocation) {
-        if(null == registry) {
-            log.warn("Please call FunctionMetricsManager.initialize() to setup metrics collection. No metrics will be pushed.");
+        if (null == registry) {
+            log.info("Please call FunctionMetricsManager.initialize() to setup metrics collection. No metrics will be pushed.");
```
[the remark this answers is not in the document]
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java:58]: Okay, I thought warn was a bit high by default. I have reverted the change.

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java (hunk 1 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:91]: Got it.

File: src/main/java/io/appform/functionmetrics/MethodData.java (hunk 3 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/MethodData.java]: Done.

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java [hunk 7]
```diff
@@ -99,60 +105,70 @@ private MethodData createMethodData(Signature callSignature) {
...
-            return Optional.empty();
-        }
-        final List<String> paramValues = IntStream.range(0, methodSignature.getMethod().getParameterCount())
-                .mapToObj(i -> {
+    private boolean isParamCaptureRequired(final MethodSignature methodSignature) {
+        return IntStream.range(0, methodSignature.getMethod().getParameterCount())
+                .anyMatch(i -> {
                     final MetricTerm metricTerm = methodSignature.getMethod()
                             .getParameters()[i].getAnnotation(MetricTerm.class);
+                    return metricTerm != null;
```
Comment (santanusinha) [inline src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:134]: inline and remove the useless variable

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java [hunk 8]
```diff
@@ -99,60 +105,70 @@ private MethodData createMethodData(Signature callSignature) {
...
+                });
+    }
+
+    private String getParamValueAtPos(final ProceedingJoinPoint joinPoint,
+                                      final int pos) {
+        final String paramValueStr = convertToString(joinPoint.getArgs()[pos]).trim();
```
Comment (santanusinha) [inline src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:140]: Array length check is needed here?

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java [hunk 9]
```diff
@@ -171,11 +187,9 @@ private void updateTimer(Timer timer, Stopwatch stopwatch) {
     private String convertToString(Object obj) {
         if (obj == null) {
             return "";
-        }
-        if (obj instanceof String) {
+        } else if (obj instanceof String) {
```
Comment (santanusinha) [inline src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:190]: else on next lines like rest of the code

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java (hunk 7 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java]: Done

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java (hunk 8 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java:141]: Done

File: src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java (hunk 9 above)
Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionTimerAspect.java]: Done

File: src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java [hunk 10]
```diff
@@ -30,16 +31,12 @@
  * Global metrics manager that needs to be initialized at start
  */
 public class FunctionMetricsManager {
-    private static final Logger log = LoggerFactory.getLogger(FunctionMetricsManager.class.getSimpleName());
+    private static final Logger log = LoggerFactory.getLogger(FunctionMetricsManager.class.getName());
```
Comment (r0goyal) [inline src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java:34]: Why this change ?

Comment (godofwharf) [reply src/main/java/io/appform/functionmetrics/FunctionMetricsManager.java:34]: To maintain the same standard of logging configuration across all the libraries. Simple name returns just "FunctionMetricsManager" whereas it should be "io.appform.functionmetrics.FunctionMetricsManager". Using the fully qualified package name would help us apply logging configuration changes in hierarchical fashion.

**Review by santanusinha — APPROVED** [review]
